File: lib/aura/ext/mcp/sse_client.rb
require "json"
require "net/http"
require "uri"
require "thread"

module Aura
  module MCP
    class SSEClient
      def initialize(url, headers = {}, timeout: 30)
        @url = URI.parse(url)
        @headers = headers || {}
        @timeout = timeout || 30
        @next_id = 1
        @queue = Queue.new
        @initialized = false
        @http = nil
        @response_queues = {}
        @lock = Mutex.new
      end

      def request(method, params = nil)
        ensure_started
        ensure_initialized
        request_raw(method, params)
      end

      def notify(method, params = nil)
        ensure_started
        payload = { "jsonrpc" => "2.0", "method" => method }
        payload["params"] = params if params
        post_message(payload)
        true
      end

      def close
        begin
          # Signal thread to stop
          @running = false
          
          # Wait for thread to finish gracefully
          if @thread && @thread.alive?
            @thread.join(2) rescue nil
            @thread.kill if @thread && @thread.alive?
          end
          
          # Close HTTP connection
          @http&.finish if @http&.active?
        rescue StandardError
          nil
        ensure
          @thread = nil
          @http = nil
        end
      end

      private
        def ensure_started
          return if @thread && @thread.alive?
          @running = true
          @thread = Thread.new { listen_loop }
          # Wait a bit for the connection to establish
          sleep 0.5 until @http && @http.active? || !@thread.alive?
        end

        def ensure_initialized
          return if @initialized
          version = defined?(Aura::VERSION) ? Aura::VERSION : "0.1.0"
          resp = request_raw("initialize", {
            "protocolVersion" => "2025-11-25",
            "capabilities" => {},
            "clientInfo" => { "name" => "aura", "version" => version }
          })
          notify("notifications/initialized", {}) if resp && resp["result"]
          @initialized = true
        end

        def request_raw(method, params = nil)
          id = @next_id
          @next_id += 1
          payload = { "jsonrpc" => "2.0", "id" => id, "method" => method }
          payload["params"] = params if params
          
          q = Queue.new
          @lock.synchronize { @response_queues[id.to_s] = q }
          
          post_message(payload)
          
          begin
            Timeout.timeout(@timeout) { q.pop }
          ensure
            @lock.synchronize { @response_queues.delete(id.to_s) }
          end
        rescue Timeout::Error
          { "error" => { "message" => "mcp sse timeout" } }
        rescue StandardError => e
          { "error" => { "message" => e.message } }
        end

        def post_message(payload)
          # In MCP SSE, the client sends POST requests to the server's message endpoint
          # The initialize response typically contains the endpoint to POST to, 
          # but for simplicity, many servers use the same URL or a relative path.
          # Here we assume the server accepts POSTs at the same URL (simplified).
          req = Net::HTTP::Post.new(@url.path)
          @headers.each { |k, v| req[k] = v }
          req.body = JSON.generate(payload)
          req["Content-Type"] = "application/json"
          
          Net::HTTP.start(@url.host, @url.port, use_ssl: @url.scheme == 'https') do |http|
            http.request(req)
          end
        rescue StandardError
          nil
        end

        def listen_loop
          Net::HTTP.start(@url.host, @url.port, use_ssl: @url.scheme == 'https') do |http|
            @http = http
            req = Net::HTTP::Get.new(@url.path)
            @headers.each { |k, v| req[k] = v }
            req["Accept"] = "text/event-stream"
            
            http.request(req) do |response|
              buffer = ""
              response.read_body do |chunk|
                break unless @running  # Check running flag
                buffer << chunk
                while (line = buffer.slice!(/.*\n/))
                  break unless @running  # Check running flag
                  handle_sse_line(line.strip)
                end
              end
            end
          end
        rescue StandardError => e
          # Connection closed or error
        end

        def handle_sse_line(line)
          return if line.empty?
          if line.start_with?("data:")
            data = line["data:".length..-1].strip
            begin
              msg = JSON.parse(data)
              handle_message(msg)
            rescue StandardError
            end
          end
        end

        def handle_message(msg)
          return unless msg.is_a?(Hash) && msg["id"]
          id = msg["id"].to_s
          q = @lock.synchronize { @response_queues[id] }
          q.push(msg) if q
        end
    end
  end
end

File: src/core/ext/mcp/sseClient.ts
import { VERSION } from '../../../utils/version.js';

export class SseClient {
  private url: string;
  private headers: Record<string, string>;
  private timeout: number;
  private nextId = 1;
  private initialized = false;

  private running = false;
  private abort: AbortController | null = null;
  private listenPromise: Promise<void> | null = null;

  private handlers = new Map<
    string,
    {
      resolve: (val: Record<string, unknown>) => void;
      reject: (err: Error) => void;
      timer: NodeJS.Timeout;
    }
  >();

  constructor(url: string, headers: Record<string, string> = {}, timeout = 30) {
    this.url = String(url);
    this.headers = headers || {};
    this.timeout = timeout || 30;
  }

  public async request(
    method: string,
    params?: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    await this.ensureStarted();
    await this.ensureInitialized();
    return this.requestRaw(method, params);
  }

  public async notify(
    method: string,
    params?: Record<string, unknown>,
  ): Promise<void> {
    await this.ensureStarted();
    await this.postMessage({ jsonrpc: '2.0', method, params });
  }

  public close(): void {
    this.running = false;
    try {
      this.abort?.abort();
    } catch (_e) {}
    this.abort = null;
    this.cleanup(new Error('MCP client closed'));
  }

  private async ensureStarted(): Promise<void> {
    if (this.running && this.listenPromise) return;

    this.running = true;
    this.abort = new AbortController();
    this.listenPromise = this.listenLoop(this.abort.signal).catch((err) => {
      this.cleanup(err instanceof Error ? err : new Error(String(err)));
    });
  }

  private async ensureInitialized(): Promise<void> {
    if (this.initialized) return;
    this.initialized = true;
    const version = VERSION;
    try {
      const resp = await this.requestRaw('initialize', {
        protocolVersion: '2025-11-25',
        capabilities: {},
        clientInfo: { name: 'aura', version },
      });
      if (resp?.result) {
        await this.postMessage({
          jsonrpc: '2.0',
          method: 'notifications/initialized',
          params: {},
        });
      }
    } catch (e) {
      this.initialized = false;
      throw e;
    }
  }

  private requestRaw(
    method: string,
    params?: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    return new Promise((resolve, reject) => {
      const id = String(this.nextId++);
      const payload = { jsonrpc: '2.0', id, method, params };

      const timer = setTimeout(() => {
        const handler = this.handlers.get(id);
        if (handler) {
          this.handlers.delete(id);
          reject(new Error('mcp sse timeout'));
        }
      }, this.timeout * 1000);

      this.handlers.set(id, { resolve, reject, timer });

      this.postMessage(payload).catch((err) => {
        clearTimeout(timer);
        this.handlers.delete(id);
        reject(err);
      });
    });
  }

  private async postMessage(payload: Record<string, unknown>): Promise<void> {
    const headers: Record<string, string> = {
      ...this.headers,
      'Content-Type': 'application/json',
    };
    await fetch(this.url, {
      method: 'POST',
      headers,
      body: JSON.stringify(payload),
    });
  }

  private async listenLoop(signal: AbortSignal): Promise<void> {
    const headers: Record<string, string> = {
      ...this.headers,
      Accept: 'text/event-stream',
    };
    const resp = await fetch(this.url, { method: 'GET', headers, signal });
    if (!resp.ok || !resp.body) {
      throw new Error(`mcp sse connect failed: ${resp.status}`);
    }

    const reader = resp.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';

    while (this.running) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      const lines = buffer.split('\n');
      buffer = lines.pop() ?? '';

      for (const raw of lines) {
        if (!this.running) break;
        const line = raw.trim();
        if (!line) continue;
        if (!line.startsWith('data:')) continue;
        const data = line.slice('data:'.length).trim();
        if (!data) continue;
        try {
          const msg = JSON.parse(data);
          this.handleMessage(msg);
        } catch (_e) {}
      }
    }
  }

  private handleMessage(msg: Record<string, unknown>): void {
    if (!msg || msg.id === undefined || msg.id === null) return;
    const id = String(msg.id);
    const handler = this.handlers.get(id);
    if (!handler) return;
    clearTimeout(handler.timer);
    this.handlers.delete(id);
    handler.resolve(msg);
  }

  private cleanup(err: Error): void {
    for (const [_, handler] of this.handlers.entries()) {
      clearTimeout(handler.timer);
      handler.reject(err);
    }
    this.handlers.clear();
    this.running = false;
    this.listenPromise = null;
    this.initialized = false;
    this.abort = null;
  }
}
